c4_preload_qsos.m

```diff
@@ -0,0 +1,91 @@
+% preload_qsos: loads spectra from SDSS FITS files, applies further
+% filters, and applies some basic preprocessing such as normalization
+% and truncation to the region of interest
+
+% load QSO catalog
+variables_to_load = {'all_QSO_ID','all_zqso', 'all_bal_flags'};
+load(sprintf('%s/catalog', processed_directory(release)), ...
+    variables_to_load{:});
+
+num_quasars = numel(all_zqso);
+
+all_wavelengths    =  cell(num_quasars, 1);
+all_flux           =  cell(num_quasars, 1);
+all_noise_variance =  cell(num_quasars, 1);
+all_pixel_mask     =  cell(num_quasars, 1);
+all_normalizers    = zeros(num_quasars, 1);
+
+% building MJD, Plate, and Fiber_ID from all_QSO_ID
+all_plates    = zeros(num_quasars,1);
+all_mjds      = zeros(num_quasars,1);
+all_fiber_ids = zeros(num_quasars,1);
+for i=1:num_quasars
+    mpf = all_QSO_ID{i};
+    all_mjds(i) = str2double(mpf(1:5));
+    all_plates(i)      = str2double(mpf(7:10));
+    all_fiber_ids(i) = str2double(mpf(12:end));
+end
+
+for i = 1:num_quasars
+
+  if (filter_flags(i)~=0)
+    continue;
+  end
+  
+  
+  [this_wavelengths, this_flux, this_noise_variance, this_pixel_mask] ...
+      = file_loader(all_mjds(i), all_plates(i),all_fiber_ids(i));
+
+  % normalize flux
+  
+  ind = (this_wavelengths >= (min_lambda * (z_qso_cut) + 1)) & ...
+        (this_wavelengths <= (max_lambda * (z_qso_training_max_cut) + 1)) & ...
+        (~this_pixel_mask);
+
+  this_median = nanmedian(this_flux(ind));
+  
+  % bit 2: cannot normalize (all normalizing pixels are masked)
+  if (isnan(this_median))
+    filter_flags(i) = bitset(filter_flags(i), 3, true);
+    continue;
+  end
+
+  % bit 3: not enough pixels available
+  if (nnz(ind) < min_num_pixels)
+    filter_flags(i) = bitset(filter_flags(i), 4, true);
+    continue;
+  end
+
+  all_normalizers(i) = this_median;
+
+  this_flux           = this_flux           / this_median;
+  this_noise_variance = this_noise_variance / this_median^2;
+
+  ind = (this_wavelengths >= loading_min_lambda) & ...
+        (this_wavelengths <= loading_max_lambda);
+        
+  % add one pixel on either side
+  available_ind = find(~ind & ~this_pixel_mask);
+  ind(min(available_ind(available_ind > find(ind, 1, 'last' )))) = true;
+  ind(max(available_ind(available_ind < find(ind, 1, 'first')))) = true;
```

Comment:
These are unused could be deleted; otherwise, should put into the next few lines.

It depends on you whether you would like to cut off the spectra.

Reply:
I don't wanna cut off the spectra, at least for the firs trial! So I will remove these lines.